Spigot-API-Patches/0058-Profile-Lookup-Events.patch

```diff
@@ -0,0 +1,221 @@
++import com.mojang.authlib.GameProfile;
++import com.mojang.authlib.GameProfileRepository;
++import com.mojang.authlib.ProfileLookupCallback;
++import com.mojang.authlib.properties.Property;
++import com.mojang.authlib.properties.PropertyMap;
++import org.bukkit.Bukkit;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++
++import javax.annotation.Nonnull;
++import javax.annotation.Nullable;
++import java.util.Set;
++import java.util.UUID;
++
++/**
++ * Allows a plugin to intercept a Profile Lookup for a Profile by name
++ *
++ * At the point of event fire, the UUID and properties are unset.
++ *
++ * If a plugin sets the UUID, and optionally the properties, the API call to look up the profile may be skipped.
++ *
++ */
++public class PreLookupProfileEvent extends Event {
++    private final String name;
++
++    private UUID uuid;
++    private PropertyMap properties = new PropertyMap();
++
++    public PreLookupProfileEvent(@Nonnull String name) {
++        super(!Bukkit.isPrimaryThread());
++        this.name = name;
++    }
++
++    /**
++     * Name of the player to look up a profile for
++     * @return
++     */
++    @Nonnull
++    public String getName() {
++        return name;
++    }
++
++    /**
++     * The UUID of the profile if it has already been provided by a plugin
++     * @return
++     */
++    @Nullable
++    public UUID getUUID() {
++        return uuid;
++    }
++
++    /**
++     * Sets the UUID for this player name. This will skip the initial API call to find the players UUID.
++     *
++     * However, if Profile Properties are needed by the server, you must also set them or else an API call will still be made.
++     * @param uuid
++     */
++    public void setUUID(@Nullable UUID uuid) {
++        this.uuid = uuid;
++    }
++
++    /**
++     * Get the properties for this profile
++     * @return
++     */
++    @Nonnull
++    public Multimap<String, Property> getProperties() {
++        return properties;
++    }
++
++    /**
++     * Completely replaces all Properties with the new provided properties
++     * @param properties
++     */
++    public void setProperties(Multimap<String, Property> properties) {
++        this.properties = new PropertyMap();
++        this.properties.putAll(properties);
++    }
++
++    /**
++     * Adds additional properties, without removing the original properties
++     * @param properties
++     */
++    public void addProperties(Multimap<String, Property> properties) {
++        this.properties.putAll(properties);
++    }
++
++    private static final HandlerList handlers = new HandlerList();
++
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++
++    /**
++     * Wraps the Profile Repository so we can intercept all lookups
++     */
++    public static GameProfileRepository wrapProfileRepository(final GameProfileRepository orig) {
```

Comment:
Probably merits more discussion, but is there a reason this has to be in the API?